Add SequenceableCollection#unixCmdGetStdOut

miguel-negrao opened the pull request:
This adds a method unixCmdGetStdOut to SequenceableCollection similar to the method in String. Internally it uses Pipe#*argv . A new primitive was added to Pipe for this. Also added documentation.

miguel-negrao commented:
AppVeyor seems to be build fine and then fail to create the zip file... The problem doesn't seem to be related to my pull request.

telephon commented on SCClassLibrary/Common/Collections/SequenceableCollection.sc:
```diff
@@ -1336,6 +1336,21 @@ SequenceableCollection : Collection {
+			line = pipe.getLine(maxLineLength);
+			while({line.notNil}, {lines = lines ++ line ++ "\n"; line = pipe.getLine; });
+			pipe.close;
+			^lines;
```
you could leave away the trailing semicolon. We are trying to do this where possible.

telephon commented on SCClassLibrary/Common/Files/File.sc:
```diff
@@ -176,6 +176,10 @@ Pipe : UnixFILE {
 	}
 
+	*argv { arg executableAndArguments, mode;
+		^super.new.openArgv(executableAndArguments, mode);
```
same here.

telephon commented on SCClassLibrary/Common/Files/File.sc:
```diff
@@ -258,4 +272,13 @@ Pipe : UnixFILE {
+			"r","w"
+		*/
+		_PipeOpenArgv
+		^this.primitiveFailed;
```
same here.

miguel-negrao commented:
@telephon removed trailing semicolon.

snappizz commented on common/sc_popen.cpp:
```diff
@@ -104,7 +104,7 @@ sc_popen_argv(const char *filename, char *const argv[], pid_t *pidp, const char
 		}
 
-		execve(filename, argv, environ);
+		execvpe(filename, argv, environ);
```
already done -- see #3501

miguel-negrao replied:
Well, sc_popen_argv had not been changed in the same way.

snappizz replied:
ah, i see.

miguel-negrao replied:
Ah, sorry, actually it had, but it is in 3.9... is that getting merged into develop soon ?

miguel-negrao commented:
Changed execve to execvp in sc_popen_argv also.

miguel-negrao commented:
> Changed execve to execvp in sc_popen_argv also.

Shouldn't have as this was already in 3.9 branch, removed that commit.

brianlheim commented:
@miguel-negrao I've messaged AppVeyor to give us a higher storage limit, this is a known problem on our end and should be resolved in a day at most.

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+	argv[0] = vfilenameOnly.data();
+	argv[obj->size] = NULL;
+
+	if(obj->size > 1) {
```
this check is unnecessary - line 1350 early-exits if this is false, and the for-loop below wouldn't run if `obj->size <= 1`

miguel-negrao replied:
ok

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+			PyrSlot argSlot;
+			getIndexedSlot(obj, &argSlot, i);
+			if (!isKindOfSlot(&argSlot, class_string)) return errWrongType;
+			char *arg = new char[slotRawObject(&argSlot)->size + 1];
```
this is a memory leak. could we just set argv[i] to the location of argSlot's char ptr, without going through an intermediary?

miguel-negrao replied:
I think I cannot use the artSlot's char ptr directly because it seems that the string is not null terminated. At least when I try to do that the debugger shows me a string which is not null terminated and slotStringVal also explicitely null- terminates the string.  slotRawObject(slot)->size  is used to know where the string ends. I will fix the memory leak though.

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+		return errWrongType;
+	if (slotRawObject(c)->size > 11) return errFailed;
+
+	PyrFile *pfile = (PyrFile*)slotRawObject(a);
```
could you use a `static_cast` here? that's more idiomatic for modern C++

miguel-negrao replied:
I tried, because cppcheck was flagging this, but it gives an error. I confess I don't understand well enough the difference between static_cast, dynamic_cast and C type cast to know why.

miguel-negrao replied:
```
error: invalid static_cast from type ‘PyrObject*’ to type ‘PyrFile*’
  PyrFile *pfile = static_cast<PyrFile*>(slotRawObject(a));
```

brianlheim replied:
ah ok, should be a reinterpret_cast then

miguel-negrao replied:
ok, changed.

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+	b = g->sp - 1; //argv
+	c = g->sp;     //mode
+
+	if (NotObj(b)) return errWrongType;
```
can you please put the `return`s on separate lines please?

miguel-negrao replied:
ok

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+
+	a = g->sp - 2;
+	b = g->sp - 1; //argv
+	c = g->sp;     //mode
```
probably better to just give these meaningful names than document their meaning in comments. `callerSlot`, `argvSlot`, `modeSlot` are good choices

miguel-negrao replied:
In other primitives that I looked at these variables are always called a,b,c,etc... Should we change that convention here ?

brianlheim replied:
Do you think it's a useful convention? I really dont.

miguel-negrao replied:
Not really...

miguel-negrao replied:
Changed to callerSlot, argvSlot, modeSlot

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+
+	PyrFile *pfile = (PyrFile*)slotRawObject(a);
+
+	PyrObject *obj = slotRawObject(b);
```
could use a more descriptive name than `obj` here, maybe `argvColl` since it's supposed to be a collection containing `argv`.

miguel-negrao replied:
ok

brianlheim commented:
Personally the variable name `executableAndArguments` seems uncommonly verbose to me. `arguments` or `args` might be more appropriate; this is what python does (https://docs.python.org/3.8/library/subprocess.html)

miguel-negrao commented:
> Thanks for this! This is just a review of the C++ additions.

Thanks for the review ! TBH, I just copy-pasted the code that was equivalent from `prArrayPOpen`. I will review the code based on your comments, and changes done based on them I can also later apply in a subsequent commit to `prArrayPOpen`.

miguel-negrao commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,74 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+	p /= filename;
+	std::string filenameOnly = p.filename().string();
+	std::vector<char> vfilenameOnly(filenameOnly.begin(), filenameOnly.end());
+	vfilenameOnly.push_back('\0');
```
Actually, i think all this code is unnecessary. It seems I was under the impression that the first element of argv should be the name of the command without full path, but actually I think it should be however the command was called in the shell, and in this case we are not calling the command from the shell so it's probably irrelevant as this is not used by programs.

miguel-negrao commented:
@brianlheim updated the pull request based on your feedback.
@snappizz already removed the commit which you commented on, so I suppose you can remove your request for changes.

Also, what is the rule for merging 3.9 into develop ? Is it ok if I do it now, to pick up the commit related to execve -> execvp ?

brianlheim commented:
> Also, what is the rule for merging 3.9 into develop ? 

I just opened #3549 to do this. I'll merge it as soon as CI passes and then you can rebase this on 3.9 or develop.

miguel-negrao commented:
> I just opened #3549 to do this. I'll merge it as soon as CI passes and then you can rebase this on 3.9 or develop (preferably 3.9, git rebase --onto 3.9 develop sequencableCollectionUnixCmdGetStdOut)

Since this is a new feature shouldn't it go into develop ?

brianlheim commented:
oops right, sorry. I got confused when you mentioned merging 3.9 in your comment above. Should go into develop :)

miguel-negrao commented:
On my side I think it is ready to merge, please review.

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+	argsSlot = g->sp - 1;
+	modeSlot = g->sp;
+
+#ifdef SC_IPHONE
```
this needs an explanatory comment

miguel-negrao replied:
Doing a git blame apparantly points to me as the author of these lines, but I have no idea why they are there, most probably someone told me that I should do this check. I imagine that on ios you can't fork ? Anyone else knows more about this ?

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+	return errNone;
+#endif
+
+	if (NotObj(argsSlot)) return errWrongType;
```
place on a separate line

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+
+	if (NotObj(modeSlot) || !isKindOf(slotRawObject(modeSlot), class_string))
+		return errWrongType;
+	if (slotRawObject(modeSlot)->size > 11) return errFailed;
```
place on a separate line

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+		getIndexedSlot(argsColl, &argSlot, i);
+		if (!isKindOfSlot(&argSlot, class_string))
+			return errWrongType;
+		char *arg = new char[slotRawObject(&argSlot)->size + 1];
```
this is still a leak if you exit early out of this for loop

brianlheim replied:
I also think you should move this loop out into a separate static method so that it's easier to understand. it could fill the vector with args and return an empty-vector on failure. this would mean separating the `new char[]` from the `delete []`.. maybe it would be safer to create a `vector<string>` which you then use to fill a `char* []`, then there is no explicit memory management required. it's fine as it is (minus memory leak) but would give you a lot of what you're going for, just more safely and with less effort

brianlheim replied:
it would also mean you can factor both these loops out of prPipeOpenArgv and prArrayPOpen!

brianlheim replied:
possible signature - `int PyrCollToStrings(PyrObject [const]* coll, std::vector<std::string>& strings)`, using the return type as an error code.

brianlheim commented on lang/LangPrimSource/PyrFilePrim.cpp:
```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
+		return errNotAnIndexableObject;
+
+	if( argsColl->size < 1)
+		return errFailed;
```
i'd appreciate it if some of these sections had explicit comments about what's being checked; my eyes glaze over when i try to read all these type checks (which are necessary, just a lot of them..)

brianlheim commented on lang/LangPrimSource/PyrUnixPrim.cpp:
```diff
@@ -186,58 +186,41 @@ int prArrayPOpen(struct VMGlobals *g, int numArgsPushed)
+	for (int i=0; i<obj->size; ++i) {
+		PyrSlot argSlot;
+		getIndexedSlot(obj, &argSlot, i);
+		if (!isKindOfSlot(&argSlot, class_string)) return errWrongType;
```
please place on a separate line

miguel-negrao commented:
@brianlheim Thanks for the new review, I will work on this and get back to you.

brianlheim commented:
When it comes time to make this cross platform the easiest thing to do would be to use Boost.Process: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_process/tutorial.html#boost_process.tutorial.env

miguel-negrao commented:
> When it comes time to make this cross platform the easiest thing to do would be to use Boost.Process: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_process/tutorial.html#boost_process.tutorial.env

The boost library seems super easy to use. I can have a look at switching to that.

miguel-negrao commented:
> When it comes time to make this cross platform the easiest thing to do would be to use Boost.Process: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_process/tutorial.html#boost_process.tutorial.env
> 
> The boost library seems super easy to use. I can have a look at switching to that.

Had a quick look. Boost:process is indeed very easy to use, but file access on the sclang side is done using normal C functions such as fgets and C file handles (FILE *). In Boost::process file reading is done with [iostreams ](http://www.boost.org/doc/libs/1_65_1/doc/html/boost/process/basic_ipstream.html). So I suppose using boost::process would require converting the whole of the file related code to c++ iostreams. Is that something we would want to do ?

brianlheim commented:
> So I suppose using boost::process would require converting the whole of the file related code to c++ iostreams. Is that something we would want to do ?

I can't speak for everyone, but YES. I think it's a losing battle trying to maintain what amounts to legacy C code. [Case in point](https://github.com/supercollider/supercollider/blob/develop/lang/LangPrimSource/PyrFilePrim.cpp#L258-L305). The unformatted IO operations on C++ iostreams are very similar to the C API. If you wanted the changes to be minimal you could use a [`basic_filebuf`](http://en.cppreference.com/w/cpp/io/basic_filebuf) pointer.

miguel-negrao commented:
@brianlheim  I've refactored the whole code in modern C++. 

I chose to use boost:variant on functions which return either a valid value or an error code. For this I had to import boost variant into the repository. I tried using the script which calls bcp but couldn't get it to work so I copied from the source of boost 1.66 by hand. If using boost variant is accepted then the variant code should be imported properly. If you give me some pointers I can try to do it again. On some functions which either fail or don't fail I used boost optional which I think is already in the repo.  We can discuss if using variant and optional is appropriate or not. 

I avoided returning values by passing in references in order to make it more evident from reading the function type what it does. For instance looking at boost::variant<int,std::string> slotStdStrVal(PyrSlot *slot) it is clear that the std::string must be created inside the function and is returned, where with passing in a parameter I don't know if the function is just altering the string or creating a new one. Boost::variant makes it clear that the function either returns an int or returns an std::string. Reading the documentation one then gathers that when it returns an int, this int is the error code.
Using optional and variant also removes bugs which can happen when code that uses the value returned tries to access it even in the case of an error. Trying the same with optional and variant immediately causes a runtime error.

I tried to make sure there were no unnecessary copies happening by using move semantics through out. I think I tested this in all the cases that mattered, verifying with the debbuger that indeed the data was being moved.

I also implemented unixCmdGetStdOut for windows but I didn't test it, as I don't have a dev setup in windows. I will use appveyor to compile for me and correct the errors afterwards.

miguel-negrao commented:
> So I suppose using boost::process would require converting the whole of the file related code to c++ iostreams. Is that something we would want to do ?
> 
> I can't speak for everyone, but YES. I think it's a losing battle trying to maintain what amounts to legacy C code. Case in point. The unformatted IO operations on C++ iostreams are very similar to the C API. If you wanted the changes to be minimal you could use a basic_filebuf pointer.

I agree with this, but I would prefer to leave refactoring all the code for file access using modern c++ to a later date. This PR would at least start to go in that direction...

miguel-negrao commented:
Ok, I figured out how to use the boost update script (with boost 1.66 sources). I only included new files, did not change files which were previously in the repo. Seems to build on windows now. I don't think I saw tests being run on windows in AppVeyor, are they ?

miguel-negrao commented:
Are pull-request's appveyor resulting binaries also uploaded to S3 ? Tried http://supercollider.s3.amazonaws.com/builds/supercollider/supercollider/win64/sequencableCollectionUnixCmdGetStdOut.html but that doesn't work.

brianlheim commented:
No, only branch builds are uploaded.

miguel-negrao commented:
Any updates on merging this ?

brianlheim commented:
I don't have time to review this now, sorry... I have to say I really didn't expect you to suddenly make these changes in the middle of this PR.

miguel-negrao commented:
> I don't have time to review this now, sorry... 

Sure, no worries.

> I have to say I really didn't expect you to suddenly make these changes in the middle of this PR.

Well, the reason for the changes was following your (very good) advice of removing all explicit memory managment, in favour of using c++'s automatic memory managment...

In the meantime it would help if someone on Windows could test this PR, for instance running:

```
["git", "status"].unixCmdGetStdOut;
["git", "status"].unixCmd;
"git status".unixCmd;
"git status".unixCmdGetStdOut;
```

If this works ok on windows then most usages of unixCmd and unixCmdGetStdOut can be switched to using the Array version, therefore avoiding the use of the shell in the Unix systems, and the problems associated with that.

brianlheim commented:
Looking it over a bit - IMO, the use of optional and variant here are kind of unidiomatic with the rest of the codebase in a way that doesn't seem to aid design.

for instance when you return a pair of pid and pointer, the pointer can already be tested for null, so wrapping the pair in an optional seems like overkill unless it's really necessary to make some future interface shift. (I can't tell if this is the case?) But testing a pointer for null is simply the idiomatic way to handle an interface that spits out pointers.

Likewise with variant - the typical usage of an error code interface is that you check the error code to determine success. Putting another layer of abstraction on top of that is counterintuitive; now we're mixing two idioms in a place where the original would work just fine. Passing the result as a pair<vector<string>, int> and constructing an empty vector is not much if at all worse than constructing a tagged union.

I think if you went with the simpler solutions, it would also require a lot less documentation on your part, because the interfaces would be self explanatory.

brianlheim commented:
> I have my own views towards how optional and variant can aid design, but since this is not even in standard C++ yet, and few people will be familiar with these types and the pattern I'm using, I will remove uses of variant and optional to avoid creating noise and distraction.
>
> In any case, let me just explain my thinking here (disclaimer, these are ideas I picked from coding in Haskell, I don't claim any originality):
> 1 - let the types document the code as much as possible.
> 2 - don't return values that don't have any defined meaning.

I'm familiar with Haskell's ADTs. But:

1. The types you've used are increasing the need for documentation here because it is an unidiomatic interface.
2. The existing return values that you replaced already have defined meaning by convention.

They might be fine if you're starting from scratch, but when there are existing conventions it just becomes noise.

The rest of what you're saying makes sense but the main issue I'm taking with it is just that we shouldn't make these kinds of changes haphazardly. Trying to force a new usage pattern like this without even discussing it first with other maintainers is counterproductive.

I have been looking at Niall Douglass's [outcome](https://ned14.github.io/outcome/) library which provides a properly named result type which is easy to use and actually designed specifically for this purpose. When we move to C++14 I think it would be good to consider whether we could rewrite some of our existing logic using result/outcome for safety and clarity.

> This then requires more documentation, not less, because I have to explain what is in that vector.

It requires the same amount of documentation - "On failure, the returned vector is empty." vs "On failure, the returned value is an error code."

miguel-negrao commented:
> I'm familiar with Haskell's ADTs. But:
> 
>     The types you've used are increasing the need for documentation here because it is an unidiomatic interface.

Agreed.

>     The existing return values that you replaced already have defined meaning by convention.

Yes, but what happens in both cases when you forget the convetion might be different. If you forget convention in the previous case and try to use the pid or try to use the empty string or vector, the error or unexpected behaviour might only manifest later, while with variant and option the error will manifest itself immediately. I have a feeling this will eventually catch more bugs. 
> 
> They might be fine if you're starting from scratch, but when there are existing conventions it just becomes noise.
> 
> The rest of what you're saying makes sense but the main issue I'm taking with it is just that we shouldn't make these kinds of changes haphazardly. Trying to force a new usage pattern like this without even discussing it first with other maintainers is counterproductive.

Ok, I can see your point there. 

I will push now my changes removing variant and option just to check if they compile on windows.

miguel-negrao commented:
It seems it is building in windows. I think this is ready for review again.

miguel-negrao commented:
Just a reminder: if this is going to be merged for 3.10 someone should test this on windows because this code is completely untested in Windows... at the moment all I know is that it compiles. Any Windows users out there willing to do that ?

brianlheim commented:
I can give it a look at some point, but my priority right now is to get the SerialPort PR through

jamshark70 commented:
If there's a test build somewhere, I can try to find a friend's machine to run it. Unfortunately, my Windows installation at school got corrupted so I don't have access to Windows on my own machines presently. I'll try to deal with that over the summer.

snappizz commented on HelpSource/Classes/Pipe.schelp:
```diff
@@ -16,9 +16,17 @@ A link::Classes/String:: representing a valid shell command.
+method::argv
+
+argument::executableAndArguments
+A link::Classes/SequenceableCollection:: containining strings where the first string is the absolute path to the executable to be run and all other strings are passed as arguments to the executable. This method starts the process directly without using a shell.
```
the "absolute path" part isn't required anymore after PATH support was added to `sc_popen_argv`, correct?

telephon commented:
ready to go?

miguel-negrao commented:
I think so...

brianlheim commented:
A whole bunch of C++ code merged in without a proper review - we should be aiming for better...

miguel-negrao commented:
> Requests have been addressed

I'm sorry, I think I misunderstood this, I thought that Brian was saying he had reviewed the code and was ok with it. It appears that was not the case, and the PR shouldn't have been merged (was this an automatic message from github when a review is dismissed ?). Sorry, it was not my intention to hurry merging the PR.

Now this is already merged, but in any case, @brianlheim if you have time in the future to have a look at the code, I can still make another PR to address any concerns that you might have at that point.

telephon commented:
> I can still make another PR to address any concerns that you might have at that point.

Yes, this is always a possibility. Obviously I also thought it is fine now after having reread the sclang code.

brianlheim commented:
I dismissed my review; did not add an approving one because I had not actually looked at it. Sorry for the confusion.

miguel-negrao commented:
> I dismissed my review; did not add an approving one because I had not actually looked at it. Sorry for the confusion.

Ok, got it. In any case thanks for the previous reviews, and when you have the time I would very much welcome a final review of the C++ code. In the mean time the code will get a bit of testing...

Future work here would be to use boost process instead of our own code. Maybe someone wants to pick that up at some point.
